src/Features/Core/Diagnostics/EngineV1/DiagnosticAnalyzerDriver.cs

```diff
@@ -10,11 +10,20 @@
 using Microsoft.CodeAnalysis.GeneratedCodeRecognition;
 using Microsoft.CodeAnalysis.Text;
 using Roslyn.Utilities;
+using System.Runtime.CompilerServices;
 
 namespace Microsoft.CodeAnalysis.Diagnostics.EngineV1
 {
     internal class DiagnosticAnalyzerDriver
     {
+        /// <summary>
+        /// Map containing unique gate objects per-analyzer instance.
+        /// We need to guard against concurrent callbacks into analyzers from different diagnostic clients.
+        /// Additionally, certain diagnostic clients, such as FixAll occurrences, may request diagnostics for multiple documents in parallel.
+        /// </summary>
+        private static readonly ConditionalWeakTable<DiagnosticAnalyzer, object> s_analyzerGateMap =
```

Comment:
That indeed seems like a better approach, but we probably want to allow future analyzers to mark themselves as concurrent or thread safe, and hence not requiring a gate. So the gate would only be needed for such analyzers. I'll let it be as is for now, but probably tweak it as per your suggestion in a future refactoring.